crates/toml/src/de.rs

```diff
@@ -45,6 +45,138 @@ where
     T::deserialize(Deserializer::new(s))
 }
 
+/// Deserializes a reader into a type.
+///
+/// This function will attempt to interpret the reader as a TOML document and
+/// deserialize `T` from the document.
+///
+/// To deserializes TOML values, instead of documents, see [`ValueDeserializer`].
+///
+/// # Examples
+///
+/// ```
+/// use serde::Deserialize;
+/// #[derive(Deserialize)]
+/// struct Config {
+///    title: String,
+///   owner: Owner,
+/// }
+///
+/// #[derive(Deserialize)]
+/// struct Owner {
+///    name: String,
+/// }
+///
+/// let config: Config = toml::from_reader(r#"
+///     title = 'TOML Example'
+///     [owner]
+///     name = 'Lisa'
+/// "#.as_bytes()).unwrap();
+///
+/// assert_eq!(config.title, "TOML Example");
+/// assert_eq!(config.owner.name, "Lisa");
+/// ```
+#[cfg(feature = "parse")]
+pub fn from_reader<R, T>(rdr: R) -> Result<T, Error>
+where
+    R: std::io::Read,
+    T: serde::de::DeserializeOwned,
+{
+    use serde::de::Error;
+
+    let mut buffer = String::new();
+    let mut rdr = rdr;
+    rdr.read_to_string(&mut buffer)
+        .map_err(crate::de::Error::custom)?;
+
+    from_str(&buffer)
+}
+
+/// Deserialize a [`Value`] into a type.
+///
+/// This function will attempt to deserialize `T` from the [`Value`].
+///
+/// # Examples
+/// ```
+/// use serde::Deserialize;
+///
+/// #[derive(Deserialize)]
+/// struct Config {
+///    title: String,
+///   owner: Owner,
+/// }
+///
+/// #[derive(Deserialize)]
+/// struct Owner {
+///     name: String,
+/// }
+///
+/// let mut table = toml::value::Table::new();
+/// table.insert("title".to_owned(), toml::value::Value::String("TOML Example".to_owned()));
+///
+/// let mut owner = toml::value::Table::new();
+/// owner.insert("name".to_owned(), toml::value::Value::String("Lisa".to_owned()));
+/// table.insert("owner".to_owned(), toml::value::Value::Table(owner));
+///
+/// let value = toml::Value::Table(table);
+///
+/// let config: Config = toml::from_value(value).unwrap();
+///
+/// assert_eq!(config.title, "TOML Example");
+/// assert_eq!(config.owner.name, "Lisa");
+/// ```
+#[cfg(feature = "parse")]
+pub fn from_value<T>(value: crate::Value) -> Result<T, Error>
```

Comment:
Can we hold off on this one?  If interested in it, I'd recommend creating an issue for us to further discuss it.

Unlike json, TOML's top-level item must always be a table.  We previously had bugs and confusing behavior because we made `Value` the top-level within the API.  So I'm a little mixed on which it should be hear and would rather we just leave it off for now

Reply:
Yes, I remove `from_value` from this PR.

Reply:
This doctest does indeed use toml::Table instead of toml::Value.

However, I don't doubt the `from_value` implementation itself.

It is just a convenience function of `serde::Deserialize::deserialize(value)`.

Currently, I have no strong incentive to move forward with this, as I do not deal directly with toml::Value.